chore: remove panics in datafusion-common::scalar by making more operations return `Result`

junjunjd opened the pull request:
## Which issue does this PR close?

It removes the majority of panics in datafusion-common::scalar https://github.com/apache/arrow-datafusion/issues/3313.

## Rationale for this change

Important move towards closing #3313
Closes #3313 

## What changes are included in this PR?

Replace most of the panics in datafusion-common::scalar by `internal_err`, `not_impl_err` or other `DataFusionError` variants

## Are these changes tested?

Yes

## Are there any user-facing changes?

No

alamb commented:
The CI appears to be failing. Marking as draft until they are passing. If there are specific questions about this PR, please let us know.

alamb commented:
Thank you for the work @junjunjd

junjunjd commented:
@alamb CI is fixed. This MR is ready for final review.

Weijun-H commented on datafusion/common/src/scalar.rs:
```diff
@@ -247,6 +247,10 @@ impl PartialEq for ScalarValue {
 }
 
 // manual implementation of `PartialOrd`
+//
+// # Panics
+//
+// Panics if there is an error when comparing kernels for arrays
```
```suggestion
```

Weijun-H replied:
Could we handle unwarp like this here 🤔
```rust
let lt_res = if let Ok(lt_res) =
    arrow::compute::kernels::cmp::lt(&arr1, &arr2)
{
    lt_res
} else {
    return None;
};

let eq_res = if let Ok(eq_res) =
    arrow::compute::kernels::cmp::eq(&arr1, &arr2)
{
    eq_res
} else {
    return None;
};
```

junjunjd replied:
Updated to remove unwrap.

Weijun-H reviewed (commented):
Thanks @junjunjd 👍

comphead reviewed (commented):
Epic work @junjunjd  thanks
I'm thinking if we can get rid of `.expect` ? Or at least provide more details in expect message?

junjunjd commented:
> Epic work @junjunjd thanks I'm thinking if we can get rid of `.expect` ? Or at least provide more details in expect message?

I removed the `.expect` in [`get_min_max_values`](https://github.com/apache/arrow-datafusion/blob/4b6180768f6ce38ab95ba1ddcc817f5bcfbd1ed7/datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs#L204) and [`get_null_count_values`](https://github.com/apache/arrow-datafusion/blob/4b6180768f6ce38ab95ba1ddcc817f5bcfbd1ed7/datafusion/core/src/datasource/physical_plan/parquet/row_groups.rs#L224) macros. 
The rest of the `.expect` calls exist in tests or examples. It makes sense to use `.expect` and panic there.

junjunjd commented:
@alamb @Weijun-H @comphead I addressed the comments. This is ready for another review.

Weijun-H commented on datafusion/common/src/scalar.rs:
```diff
@@ -3419,8 +3492,8 @@ mod tests {
 
-        let left_array = left.to_array();
-        let right_array = right.to_array();
+        let left_array = left.to_array().expect("Failed to convert to array");
```
```suggestion
        let left_array = left.to_array().expect("Failed to convert to array in add overflow check");
```

junjunjd replied:
I believe the backtrace will provide information on the test function. For example, the following context is provided on failure
`thread 'scalar::tests::scalar_add_overflow_test' panicked at datafusion/common/src/scalar.rs:3495:9:`

Weijun-H reviewed (changes requested):
Could we get rid of `.expect` in tests?

houqp commented:
@Weijun-H @comphead using unwrap and expect in tests is actually the preferred practice, see https://github.com/influxdata/influxdb/blob/main/docs/style_guide.md#dont-return-result-from-test-functions. It makes the test failure easier to parse for a human and the test framework will already provide all  the necessary context.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -3468,22 +3541,30 @@ mod tests {
 
         // decimal scalar to array
-        let array = decimal_value.to_array();
+        let array = decimal_value
+            .to_array()
+            .expect("Failed to convert to array");
```
I think it makes sense to use `.expect` here so that we can have more information on test failure. See my other comment https://github.com/apache/arrow-datafusion/pull/7901#issuecomment-1780426607

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -3468,22 +3541,30 @@ mod tests {
 
         // decimal scalar to array with size
-        let array = decimal_value.to_array_of_size(10);
+        let array = decimal_value
+            .to_array_of_size(10)
+            .expect("Failed to convert to array of size");
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -3468,22 +3541,30 @@ mod tests {
         // test eq array
-        assert!(decimal_value.eq_array(&array, 1));
-        assert!(decimal_value.eq_array(&array, 5));
+        assert!(decimal_value
+            .eq_array(&array, 1)
+            .expect("Failed to compare arrays"));
+        assert!(decimal_value
+            .eq_array(&array, 5)
+            .expect("Failed to compare arrays"));
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -3419,8 +3492,8 @@ mod tests {
-        let left_array = left.to_array();
-        let right_array = right.to_array();
+        let left_array = left.to_array().expect("Failed to convert to array");
+        let right_array = right.to_array().expect("Failed to convert to array");
```
Same as above.

Weijun-H reviewed (approved):
LGTM! Thanks @junjunjd

alamb reviewed (commented):
Thank you for this PR @junjunjd. We very much appreciate the effort -- I am sorry that the ticket description may have been misleading

Buried in https://github.com/apache/arrow-datafusion/issues/3313, it says

> The goal is not to remove all panics but review and make sure we are using them appropriately. Bonus points for adding documentation for invariants.

Can you explain why you removed the panics that you did? I think most of them are "unreachable" so forcing client code to check for errors that will never happen makes it harder to work with (and is why this PR adds around 500 new lines of code)

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -330,9 +330,9 @@ impl PartialOrd for ScalarValue {
 
                         let lt_res =
-                            arrow::compute::kernels::cmp::lt(&arr1, &arr2).unwrap();
+                            arrow::compute::kernels::cmp::lt(&arr1, &arr2).ok()?;
```
This panic is reachable, for example if `arr1` and `arr2` have different data type, [`arrow::compute::kernels::cmp::lt`](https://github.com/apache/arrow-rs/blob/master/arrow-ord/src/cmp.rs#L201-L204) will panic.

It makes sense to return `None` here instead of panicking and exiting since user just performs a partial order comparison.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -330,9 +330,9 @@ impl PartialOrd for ScalarValue {
+                            arrow::compute::kernels::cmp::lt(&arr1, &arr2).ok()?;
                         let eq_res =
-                            arrow::compute::kernels::cmp::eq(&arr1, &arr2).unwrap();
+                            arrow::compute::kernels::cmp::eq(&arr1, &arr2).ok()?;
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -512,19 +516,19 @@ impl std::hash::Hash for ScalarValue {
-    let dict_array = as_dictionary_array::<K>(array).unwrap();
-    (dict_array.values(), dict_array.key(index))
+) -> Result<(&ArrayRef, Option<usize>)> {
+    let dict_array = as_dictionary_array::<K>(array)?;
```
This is a public function and `as_dictionary_array` can fail depending on what `array` value the user passes to the function. The [`downcast_value`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/lib.rs#L73) macro called by `as_dictionary_array` returns an internal error if the cast is impossible. 

Returning a `Result` here makes sense because this error is recoverable.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -579,24 +581,44 @@ fn dict_from_values<K: ArrowDictionaryKeyType>(
+                    "could not cast value to {}",
+                    type_name::<$ARRAYTYPE>()
+                ))
+            })?;
```
`typed_cast_tz` macro is called by [`try_from_array`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L2021-L2022). `try_from_array` is a public function.  Depending on what `array` value the user passes to the function, downcasting the array to a timestamp array can fail. 

This error is reachable and recoverable. It makes sense to  return an internal error here. This aligns with how downcast error is handled in the [downcast_value](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/lib.rs#L73) macro.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -579,24 +581,44 @@ fn dict_from_values<K: ArrowDictionaryKeyType>(
+                    "could not cast value to {}",
+                    type_name::<$ARRAYTYPE>()
+                ))
+            })?;
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1309,27 +1338,33 @@ impl ScalarValue {
-                        sv => panic!(
-                            "Inconsistent types in ScalarValue::iter_to_array. \
+                            sv => _internal_err!(
+                                "Inconsistent types in ScalarValue::iter_to_array. \
```
It appears that unreachable errors are handled inconsistently in `iter_to_array`. This error should be [unreachable](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1277). However, all the other `build_*` macros defined in `iter_to_array` return an internal error (for example [`build_array_list_string`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1253), [`build_array_primitive`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1145). I removed `panic` here to align with other `build_*` macros.
This does not require any code change in client side, since `iter_to_array` already returns a Result.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1539,7 +1574,7 @@ impl ScalarValue {
                                 Ok(*scalar)
                             } else {
-                                panic!("Expected inner key type of {key_type} but found: {inner_key_type}, value was ({scalar:?})");
+                                _internal_err!("Expected inner key type of {key_type} but found: {inner_key_type}, value was ({scalar:?})")
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1562,7 +1597,11 @@ impl ScalarValue {
+                        return _internal_err!(
+                            "Invalid dictionary keys type: {key_type:?}"
+                        )
+                    }
```
I believe this error is reachable if the user passes dictionaries with non-integer keys to the function. It is recoverable so I replaced `unreachable!` with an internal error.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1611,15 +1650,19 @@ impl ScalarValue {
+            |r, element: ScalarValue| match element {
+                ScalarValue::Null => Ok::<usize, DataFusionError>(r + 1),
+                s => {
+                    _internal_err!("Expected ScalarValue::Null element. Received {s:?}")
```
Similar to the previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1378446881) I made to the `build_*` macros in iter_to_array, this error is reachable. Returning an internal error here aligns with the rest of the function.
This does not require any code change in client side, since [iter_to_array](url) already returns a Result.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1630,10 +1673,12 @@ impl ScalarValue {
+                ScalarValue::Decimal128(v1, _, _) => Ok(v1),
+                s => {
+                    _internal_err!("Expected ScalarValue::Null element. Received {s:?}")
+                }
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1646,10 +1691,14 @@ impl ScalarValue {
+                    _internal_err!(
+                        "Expected ScalarValue::Decimal256 element. Received {s:?}"
+                    )
+                }
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -579,24 +581,44 @@ fn dict_from_values<K: ArrowDictionaryKeyType>(
 
 macro_rules! typed_cast_tz {
     ($array:expr, $index:expr, $ARRAYTYPE:ident, $SCALAR:ident, $TZ:expr) => {{
-        let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
```
typed_cast_tz macro is called by [try_from_array](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L2021-L2022). try_from_array is a public function. Depending on what array value the user passes to the function, downcasting the array to a timestamp array can fail.
This error is reachable and recoverable. It makes sense to return an internal error here. This aligns with how downcast error is handled in the [downcast_value](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/lib.rs#L73) macro.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -579,24 +581,44 @@ fn dict_from_values<K: ArrowDictionaryKeyType>(
 
 macro_rules! typed_cast {
     ($array:expr, $index:expr, $ARRAYTYPE:ident, $SCALAR:ident) => {{
-        let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -676,13 +696,13 @@ macro_rules! build_values_list {
                     ScalarValue::$SCALAR_TY(None) => {
                         builder.values().append_null();
                     }
-                    _ => panic!("Incompatible ScalarValue for list"),
```
I agree this error can be unreachable if invariant is established that the `data_type` and `values` arguments passed to [`new_list`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1664) align with each other. 
I will change this back to `panic` and thus making `new_list` panic instead of returning `Result`.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -701,13 +721,13 @@ macro_rules! build_values_list_tz {
                         builder.values().append_null();
                     }
-                    _ => panic!("Incompatible ScalarValue for list"),
+                    _ => return _internal_err!("Incompatible ScalarValue for list"),
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -739,12 +759,21 @@ macro_rules! build_timestamp_array_from_option {
 
 macro_rules! eq_array_primitive {
     ($array:expr, $index:expr, $ARRAYTYPE:ident, $VALUE:expr) => {{
-        let array = $array.as_any().downcast_ref::<$ARRAYTYPE>().unwrap();
```
This error can be unreachable if an invariant is established that the `array` passed to [`eq_array`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1664) always has the same data type, precision and scale with `self`.
I will change this back to panic and thus making `eq_array` panic instead of returning Result.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1562,7 +1597,11 @@ impl ScalarValue {
                     DataType::UInt16 => dict_from_values::<UInt16Type>(values)?,
                     DataType::UInt32 => dict_from_values::<UInt32Type>(values)?,
                     DataType::UInt64 => dict_from_values::<UInt64Type>(values)?,
-                    _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
```
I believe this error is reachable if the user passes dictionaries with non-integer keys to the function. It is recoverable so I replaced unreachable! with an internal error.
This does not require any code change in client side, since `iter_to_array` already returns a Result.

junjunjd replied:
I agree it makes more sense to keep `unreachable!` here to align with how `arrow-rs` deals with illegal dictionary key types.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1748,17 +1797,17 @@ impl ScalarValue {
         match value {
             Some(val) => Decimal128Array::from(vec![val; size])
                 .with_precision_and_scale(precision, scale)
-                .unwrap(),
```
Since [`ScalarValue::try_new_decimal128`](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L660-L661) allows a decimal with precision 0, this error is reachable when user creates a decimal with precision 0 and [`arrow-array`](https://github.com/apache/arrow-rs/blob/master/arrow-array/src/types.rs#L1234-L1235) does not support 0 precision decimal.
Would it make sense to disallow a decimal with precision 0 in `try_new_decimal128` so that this error becomes unreachable?

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1748,17 +1797,17 @@ impl ScalarValue {
             None => {
                 let mut builder = Decimal128Array::builder(size)
                     .with_precision_and_scale(precision, scale)
-                    .unwrap();
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1768,12 +1817,12 @@ impl ScalarValue {
             .take(size)
             .collect::<Decimal256Array>()
             .with_precision_and_scale(precision, scale)
-            .unwrap()
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1834,30 +1883,31 @@ impl ScalarValue {
 
                 let arr = Decimal128Array::from(vals)
-                    .with_precision_and_scale(*precision, *scale)
-                    .unwrap();
```
Similar to previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381174929), this error is reachable because [ScalarValue::try_new_decimal128](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L660-L661) allows a decimal with precision 0, while [arrow-array](https://github.com/apache/arrow-rs/blob/master/arrow-array/src/types.rs#L1234-L1235) does not support that.
We can update `try_new_decimal128` to disallow decimal with precision 0 or establish an invariant here so this error becomes unreachable and we can panic here.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2089,19 +2140,23 @@ impl ScalarValue {
-                    DataType::UInt16 => dict_from_scalar::<UInt16Type>(v, size),
-                    DataType::UInt32 => dict_from_scalar::<UInt32Type>(v, size),
-                    DataType::UInt64 => dict_from_scalar::<UInt64Type>(v, size),
-                    _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
```
See previous `[comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381195593)`.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2259,71 +2316,63 @@ impl ScalarValue {
-                    DataType::UInt16 => get_dict_value::<UInt16Type>(array, index),
-                    DataType::UInt32 => get_dict_value::<UInt32Type>(array, index),
-                    DataType::UInt64 => get_dict_value::<UInt64Type>(array, index),
-                    _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
```
See previous `[comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381195593)`.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2481,119 +2529,136 @@ impl ScalarValue {
-                    DataType::UInt16 => get_dict_value::<UInt16Type>(array, index),
-                    DataType::UInt32 => get_dict_value::<UInt32Type>(array, index),
-                    DataType::UInt64 => get_dict_value::<UInt64Type>(array, index),
-                    _ => unreachable!("Invalid dictionary keys type: {:?}", key_type),
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381195593).

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1801,29 +1850,29 @@ impl ScalarValue {
                 build_timestamp_list!(unit.clone(), tz.clone(), values, 1)
             }
             DataType::List(_) | DataType::Struct(_) => {
-                ScalarValue::iter_to_array_list_without_nulls(values, data_type).unwrap()
```
Similar to previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381174929), this error is reachable if `values` are decimals with precision 0.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1834,30 +1883,31 @@ impl ScalarValue {
-            _ => panic!(
-                "Unsupported data type {:?} for ScalarValue::list_to_array",
-                data_type
-            ),
```
By updating `try_new_decimal128` to disallow decimal with precision 0 and establishing some invariants, errors in `new_list` should be unreachable. I will change `new_list` back to panic to avoid impact on client code.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1834,30 +1883,31 @@ impl ScalarValue {
+    pub fn to_array_of_size(&self, size: usize) -> Result<ArrayRef> {
+        Ok(match self {
             ScalarValue::Decimal128(e, precision, scale) => Arc::new(
-                ScalarValue::build_decimal_array(*e, *precision, *scale, size),
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381174929).

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1947,14 +1997,14 @@ impl ScalarValue {
                         repeat(None::<&[u8]>).take(size),
                         *s,
                     )
-                    .unwrap(),
```
Same as above.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1834,30 +1883,31 @@ impl ScalarValue {
+                ScalarValue::build_decimal_array(*e, *precision, *scale, size)?,
             ),
             ScalarValue::Decimal256(e, precision, scale) => Arc::new(
-                ScalarValue::build_decimal256_array(*e, *precision, *scale, size),
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381174929).

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1970,13 +2020,14 @@ impl ScalarValue {
                 ),
             },
             ScalarValue::Fixedsizelist(..) => {
-                unimplemented!("FixedSizeList is not supported yet")
```
What would be the preferred way to handle unimplemented errors in `datafusion`? There are many places where a `NotImplemented` error is returned instead of using `unimplemented!` and panicking. IMO returning an error makes more sense as user can choose to ignore unimplemented errors instead of panicking and exiting.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -1970,13 +2020,14 @@ impl ScalarValue {
                 let arrays = std::iter::repeat(arr.as_ref())
                     .take(size)
                     .collect::<Vec<_>>();
-                arrow::compute::concat(arrays.as_slice()).unwrap()
```
I agree this should be unreachable. I will change this back to unwrap.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2350,7 +2399,9 @@ impl ScalarValue {
                 let array = as_fixed_size_binary_array(array)?;
                 let size = match array.data_type() {
                     DataType::FixedSizeBinary(size) => *size,
-                    _ => unreachable!(),
```
I agree this should be unreachable. I will change this back to unwrap.

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2462,17 +2511,16 @@ impl ScalarValue {
                     *precision,
                     *scale,
-                )
-                .unwrap()
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1379698814).

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2481,119 +2529,136 @@ impl ScalarValue {
                     *precision,
                     *scale,
-                )
-                .unwrap()
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1379698814).

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2481,119 +2529,136 @@ impl ScalarValue {
+                eq_array_primitive!(array, index, LargeBinaryArray, val)?
             }
-            ScalarValue::Fixedsizelist(..) => unimplemented!(),
-            ScalarValue::List(_) => unimplemented!("ListArr"),
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381231326).

junjunjd commented on datafusion/common/src/scalar.rs:
```diff
@@ -2481,119 +2529,136 @@ impl ScalarValue {
+            ScalarValue::Struct(_, _) => {
+                return _not_impl_err!("Struct is not supported yet")
             }
-            ScalarValue::Struct(_, _) => unimplemented!(),
```
See previous [comment](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381231326).

junjunjd commented:
@alamb Thanks for the review. I have added comments to the panics I removed in [scalar.rs](https://github.com/apache/arrow-datafusion/pull/7901/files#diff-3ed49bf2d745740a87441ae1fcdec44fe853f4809ee671fe294818982e020da6). To summarize, these panics can be categorized into five types in general:

- [panics](https://github.com/apache/arrow-datafusion/pull/7901/files#r1378446881) generated in `iter_to_array` when the `ScalarValue`s in the iterator [are not all the same type](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1093-L1094). I believe these errors are reachable. Most of the [`build_*` macros](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L1145) defined in the function return an internal error instead of panicking. IMO it makes sense to remove these panics to align with other internal errors returned.
- [`typed_cast_*` macros](https://github.com/apache/arrow-datafusion/pull/7901/files#r1379681340) called in `try_from_array`. Since `try_from_array` is a public function, downcasting the array to certain types can fail depending on what `array` value the user passes to the function.
I think it makes sense to return an internal error as this error is reachable and recoverable. This aligns with how downcast error is handled in the [downcast_value](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/lib.rs#L73) macro. `try_from_array` already returns a `Result`, so this does not require any change in client code.
- [panics](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381174929) generated when `with_precision_and_scale` is called on decimal arrays. I think these errors are reachable because [ScalarValue::try_new_decimal128](https://github.com/apache/arrow-datafusion/blob/main/datafusion/common/src/scalar.rs#L660-L661) allows decimals with precision 0 while [arrow-array](https://github.com/apache/arrow-rs/blob/master/arrow-array/src/types.rs#L1234-L1235) does not support that. We can update `try_new_decimal128` to disallow decimal with precision 0 and establish some other invariants to [`new_list`](https://github.com/apache/arrow-datafusion/pull/7901/files#r1379688238), [`eq_array`](https://github.com/apache/arrow-datafusion/pull/7901/files#r1379698814) and [`to_array_of_size`](https://github.com/apache/arrow-datafusion/pull/7901/files#r1381221881) so that these error becomes unreachable and these functions can panic. This should remove the impacts on client code.
- the unimplemented errors. In many other `datafusion` code, a `NotImplemented` error is returned instead of using `unimplemented!`. IMHO returning an error makes more sense as user can choose whether to ignore the unimplemented errors instead of panicking and exiting. Would appreciate your thoughts on this.
- All the `ArrowError`s and the "Invalid dictionary keys type" errors should be unreachable. I will change these back to panics.

alamb commented:
@junjunjd  can you please merge and resolve the conflicts in this PR? Then we can merge it in

alamb commented:
Marking as draft as it needs conflicts resolved prior to being mergable.

junjunjd commented:
@alamb Thank you for the review! I rebased the MR. It is ready for final review/merge.

alamb reviewed (approved):
Thank you @junjunjd
